rust/connlib/tunnel/src/client.rs

```diff
@@ -151,18 +152,56 @@ pub struct ClientState {
 }
 
 enum DnsResourceNatState {
-    Pending { sent_at: Instant },
+    Pending {
+        sent_at: Instant,
+        buffered_packets: AllocRingBuffer<IpPacket>,
+    },
     Confirmed,
 }
 
 impl DnsResourceNatState {
-    fn confirm(&mut self) {
-        *self = Self::Confirmed;
+    fn num_buffered_packets(&self) -> usize {
+        match self {
+            DnsResourceNatState::Pending {
+                buffered_packets, ..
+            } => buffered_packets.len(),
+            DnsResourceNatState::Confirmed => 0,
+        }
+    }
+
+    fn confirm(&mut self) -> impl Iterator<Item = IpPacket> {
+        let buffered_packets = match std::mem::replace(self, DnsResourceNatState::Confirmed) {
+            DnsResourceNatState::Pending {
+                buffered_packets, ..
+            } => Some(buffered_packets.into_iter()),
+            DnsResourceNatState::Confirmed => None,
+        };
+
+        buffered_packets.into_iter().flatten()
     }
 }
 
 struct PendingFlow {
     last_intent_sent_at: Instant,
+    packets: AllocRingBuffer<IpPacket>,
+}
+
+impl PendingFlow {
+    /// How many packets we will at most buffer in a [`PendingFlow`].
+    ///
+    /// `PendingFlow`s are per _resource_ (which could be Internet Resource or wildcard DNS resources).
+    /// Thus, we may receive a fair few packets before we can send them.
+    const CAPACITY_POW_2: usize = 7; // 2^7 = 128
```

Comment:
Are we buffering UDP too or just ICMP / TCP?

If the former, would it make sense for this to be more than 128, like 1024?

Just wondering if certain write-heavy UDP protocols might see a benefit from this.